stablehlo/transforms/StablehloRefineShapes.cpp

```diff
@@ -577,19 +578,48 @@ struct EvalSliceOpPattern : public OpRewritePattern<SliceOp> {
   LogicalResult matchAndRewrite(SliceOp op,
                                 PatternRewriter& rewriter) const override {
     auto resultType = op.getType();
-    if (!resultType.hasRank() || resultType.getRank() != 1)
-      return rewriter.notifyMatchFailure(op, "expected 1-dimensional type");
+    if (!resultType.hasRank() || resultType.getRank() < 1)
+      return rewriter.notifyMatchFailure(
+          op, "expected non-0 ranked tensor result type");
+
+    auto operand = op.getOperand().dyn_cast<TypedValue<RankedTensorType>>();
+    if (!operand || !operand.getType().hasStaticShape())
+      return rewriter.notifyMatchFailure(
+          op, "expected operand with static ranked tensor type");
+
+    // A ranked tensor type with unit dimension prefix of R-1 size is physically
+    // compatible with 1-dimensional type.
+    auto resultShape = resultType.getShape();
+    for (const auto& dimSize : resultShape.drop_back(1)) {
+      if (dimSize != 1)
+        return rewriter.notifyMatchFailure(
+            op, "expected 1-dimensional result type");
+    }
 
-    SmallVector<APSInt> operand;
-    if (failed(hlo::matchInts(op.getOperand(), operand)))
+    SmallVector<APSInt> operandData;
+    if (failed(hlo::matchInts(op.getOperand(), operandData)))
       return rewriter.notifyMatchFailure(op, "expected constant operand");
 
-    int64_t start = op.getStartIndices()[0];
-    int64_t limit = op.getLimitIndices()[0];
-    int64_t stride = op.getStrides()[0];
+    assert(llvm::all_of(op.getStrides().drop_back(),
+                        [](int64_t s) { return s == 1; }) &&
+           "all strides but last must be 1");
```

Comment:
I need to revisit this ops verifier to see if this is guaranteed. If this is a part of the op verifier (dimSize 1 in result shape must have stride 1) then I don't think we need this assert. If this isn't verified, then I think asserting may lead to some crashes here.

Reply:
That is also my concern expressed in this assert, wanted to restrict possible cases. But i assume that writing something like this:
```
%1 = "stablehlo.slice"(%0) {
    start_indices = array<i64: 0, 0>,
    limit_indices = array<i64: 1, 1>,
    strides = array<i64: 99, 42>
  } : (tensor<1x3xi64>) -> tensor<1x1xi64>
```
is probably valid as whatever stride is it never does 2nd iteration. Therefore assert should be removed?

Reply:
Looks like we only use stride to enforce the result shape:

`shape(result) = ceil((limit_indices - start_indices) / strides)`

As you point out, the following is a valid program that we would (maybe?) want to fold:

```
module {
  func.func @slice_test(%arg0: tensor<1x3xi64>) -> tensor<1x1xi64> {
    %0 = stablehlo.slice %arg0 [0:1:99, 0:1:42] : (tensor<1x3xi64>) -> tensor<1x1xi64>
    return %0 : tensor<1x1xi64>
  }
}
```

With that said I'm ok with either:

1. Support folding the above program: Remove this assert.
2. Don't support folding this program: Replace the assert with a `return notifyMatchFailure`

Will leave the decision up to you, 2 is OK since it is non-standard IR. We don't _need_ to support this program until there is a use case for it. Regardless we don't want to assert/crash on valid IR that isn't supported by this pass